test/expr/delayed-ident/member_chains.swift

```diff
@@ -0,0 +1,211 @@
+let _: ImplicitMembers = .implicit.another[optional: ()]!
+let _: ImplicitMembers? = .implicit.another[optional: ()]
+let _: ImplicitMembers = .implicit.another[func: ()]()
+let _: ImplicitMembers = .implicit.another[funcOptional: ()]()!
+let _: ImplicitMembers? = .implicit.another[funcOptional: ()]()
+let _: ImplicitMembers = .implicit.another[optionalFunc: ()]!()
+let _: ImplicitMembers? = .implicit.another[optionalFunc: ()]?()
+let _: ImplicitMembers = .implicit[()].another
+let _: ImplicitMembers = .implicit[optional: ()]!.another
+let _: ImplicitMembers? = .implicit[optional: ()]?.another
+let _: ImplicitMembers = .implicit[func: ()]().another
+let _: ImplicitMembers = .implicit[funcOptional: ()]()!.another
+let _: ImplicitMembers? = .implicit[funcOptional: ()]()?.another
+let _: ImplicitMembers = .implicit[optionalFunc: ()]!().another
+let _: ImplicitMembers? = .implicit[optionalFunc: ()]?().another
+
+func implicit(_ i: inout ImplicitMembers) {
+    if i == .implicit {}
+    if i == .implicit.another {}
+    if i == .implicit.getAnother() {}
+    if i == .optional?.another {}
+    if i == .optional!.another {}
+    if i == .createOptional()?.another {}
+}
+
+struct ImplicitGeneric<T> {
+    static var implicit: ImplicitGeneric<T> { ImplicitGeneric<T>() }
+    var another: ImplicitGeneric<T> { ImplicitGeneric<T>() }
+    func getAnother() -> ImplicitGeneric<T> {
+        ImplicitGeneric<T>()
+    }
+}
+
+extension ImplicitGeneric where T == Int {
+    static var implicitInt: ImplicitGeneric<Int> { ImplicitGeneric<Int>() }
+    static var implicitString: ImplicitGeneric<String> { ImplicitGeneric<String>() }
+    var anotherInt: ImplicitGeneric<Int> { ImplicitGeneric<Int>() }
+    var anotherIntString: ImplicitGeneric<String> { ImplicitGeneric<String>() }
+    func getAnotherInt() -> ImplicitGeneric<Int> {
+        ImplicitGeneric<Int>()
+    }
+}
+
+extension ImplicitGeneric where T == String {
+    static var implicitString: ImplicitGeneric<String> { ImplicitGeneric<String>() }
+    var anotherString: ImplicitGeneric<String> { ImplicitGeneric<String>() }
+    var anotherStringInt: ImplicitGeneric<Int> { ImplicitGeneric<Int>() }
+    func getAnotherString() -> ImplicitGeneric<String> {
+        ImplicitGeneric<String>()
+    }
+}
+
+func implicit<T>(_ arg: ImplicitGeneric<T>) {}
+
+implicit(.implicitInt)
+implicit(.implicit.anotherInt)
+implicit(.implicit.anotherInt.another)
+implicit(.implicit.another.anotherInt)
+implicit(.implicit.getAnotherInt())
+implicit(.implicit.another.getAnotherInt())
+implicit(.implicit.getAnother().anotherInt)
+implicit(.implicit.getAnotherInt())
+implicit(.implicit.getAnother().getAnotherInt())
+implicit(.implicitString.anotherStringInt)
+// Member types along the chain can have different generic arguments
+implicit(.implicit.anotherIntString.anotherStringInt)
+
+implicit(.implicit.anotherString.anotherStringInt) // expected-error {{type of expression is ambiguous without more context}}
```

Comment:
Not sure why 

```
implicit(.implicit.anotherIntString.anotherStringInt)
```

compiles but

```
implicit(.implicit.anotherString.anotherStringInt)
```

doesn't. They seem to be testing very similar behavior... 🤔 (Also, the latter seems to be duplicated on line 210.)

Reply:
The difference is a bit subtle, but if you look at the final member of each chain, they both produce `ImplicitGeneric<Int>`, which implies that the base of the chain should also be `ImplicitGeneric<Int>`. Then, `.implicit` produces `ImplicitGeneric<Int>`, which *does* have `anotherIntString` as a member but does not have `anotherString` as a member (since that’s only a member on `ImplicitGeneric<String>`.

Per a conversation in the forum thread, I’m looking at changing the implementation to allow chains that differ only in generic parameters in the head and tail, which should allow this to compile as well. At the very least, the diagnostic here should be improved.

Reply:
Thanks, that makes sense. But that raises another question: how is this backwards reasoning even working 😅?

> but if you look at the final member of each chain, they both produce ImplicitGeneric<Int>, which implies that the base of the chain should also be ImplicitGeneric<Int>.

To know the type of the final member in the chain, don't you need to know the type of the first member and then follow the chain? (I'm guessing this is related to Robert's point about inside-out vs outside-in) Is it that because of the context, we know there's going to be some `ImplicitGeneric<T>`, then we look at the final member in the chain -> figure out `T = Int` (probably by looking at all the properties of `ImplicitGeneric<T>` including those in extensions), then we take the base of the chain to be `ImplicitGeneric<Int>`?

Reply:
Yeah, that's pretty much how it works. The basic concept that makes this possible is that lookup doesn't require generic parameters to be resolved in order to find members, so once the context is used to resolve the base to `ImplicitMembers<_>`, we can perform lookup along the chain and resolve the generic parameters based on the members themselves.

In the single-member chain situation, this already works. To pull an example that Joe gave in the [discussion thread](https://forums.swift.org/t/allow-chained-member-references-in-implicit-member-expressions/32829/12), the following code already typechecks:
```
struct Foo<T> {}

extension Foo where T == Int { static var x = Foo<Int>() }

func foo<T>(_: Foo<T>) { print(T.self) }

foo(.x)
```

Reply:
Thanks, appreciate the explanation!

Reply:
Np!